doc/1.4/language.md

```diff
@@ -3300,29 +3355,156 @@ after h -> (x, y): my_callback(i, y + 1.5, false)
 ```
 as the message component parameter `y` is used, but not as a direct argument.
 
-For both forms of `after`, to allow the suspended method call to be represented
-in a checkpoint, `after` statements may only be performed with methods that have
-no return values, and where each input parameter of the method must be of
-[*serializable type*](#serializable-types), unless that input parameter receives
-a message component.
+#### Immediate After Statements
+<pre>
+after: <em>method</em>(<em>e1</em>, ... <em>eN</em>);
+</pre>
 
-This means that `after` statements cannot be used with methods that e.g. have
-pointer input parameters, unless the arguments for those input parameters are
-message component parameters of the `after`.
+In this form, the specified point in the future when control is given back to
+the simulation engine such that the ongoing simulation of the current processor
+may progress, and the simulation would otherwise be ready to move onto the next
+cycle.
+This happens after all entries to devices on the call stack have been completed.
+
+Immediate after statements are most useful to avoid ordering bugs. It can be
+used to delay a method call until all current lines of execution into the device
+have been completed, and the device is guaranteed to be in a consistent state
+where it is ready to handle the method call.
+
+Semantically, the immediate after statement is very close to
+`after 0 cycles: ...`, but has a number of advantages. In general, the immediate
+after statement is designed to execute the callback as promptly as possible
+while satisfying the above conditions, while `after 0 cycles: ...` is not.
+In particular, in Simics, callbacks delayed via `after 0 cycles` are always
+bound to the clock associated with the device instance, which is not always
+that of the processor currently under simulation &mdash; in such cases the
+simulated processor may progress indefinitely without the posted callback being
+executed. In addition, if an immediate after statement is executed while the
```

Comment:
Clarification: "... without the posted callback being executed. The immediate after statement does not have this issue."